restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/CallManager.java

```diff
@@ -649,6 +655,152 @@ private void info(final SipServletRequest request) throws IOException {
         }
     }
 
+    private void transfer(SipServletRequest request) throws Exception {
+        final SipApplicationSession appSession = request.getApplicationSession();
+        ActorRef call = (ActorRef) appSession.getAttribute(Call.class.getName());
+        if (call == null) {
+            SipServletResponse servletResponse = request.createResponse(SC_TEMPORARILY_UNAVAILABLE, "REFER should be sent in dialog");
+            servletResponse.setHeader("Event", "refer");
+            servletResponse.setHeader("Retry-After", "1800000");
+            servletResponse.send();
+            return;
+        }
+
+        final Timeout expires = new Timeout(Duration.create(60, TimeUnit.SECONDS));
+
+        Future<Object> infoFuture = (Future<Object>) ask(call, new GetCallInfo(), expires);
+        CallResponse<CallInfo> infoResponse = (CallResponse<CallInfo>) Await.result(infoFuture,
+                Duration.create(10, TimeUnit.SECONDS));
+        CallInfo callInfo = infoResponse.get();
+
+        CallDetailRecordsDao dao = storage.getCallDetailRecordsDao();
+
+        CallDetailRecord cdr = dao.getCallDetailRecord(callInfo.sid());
+        Sid parentCallSid = cdr.getParentCallSid();
+        cdr = dao.getCallDetailRecord(parentCallSid);
+
+        call = lookup(new GetCall(cdr.getCallPath()));
+        IncomingPhoneNumber number = storage.getIncomingPhoneNumbersDao().getIncomingPhoneNumber(cdr.getTo());
+
+        if (number.getReferUrl() == null && number.getReferApplicationSid() == null) {
+            SipServletResponse servletResponse = request.createResponse(SC_TEMPORARILY_UNAVAILABLE, "Set either incoming phone number Refer URL or Refer application");
```

Comment:
@vmykhailiuk I think that if referUrl is null we should response with a more appropriate error such as 404 or 488.